### PR TITLE
Feat: Add `movedim` tensor operator

### PR DESCRIPTION
**Pull request by LilDojd**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

None

### Changes

Added `movedim` default implementation. Mirrors the [PyTorch one](https://pytorch.org/docs/stable/generated/torch.movedim.html).

This just wraps a call to permute with sanity checks, but moving one/few axes while keeping all others inbound is a common enough operation to warrant its specific convenient syntactic sugar.

### Testing

Added unit tests.

### CODE REVIEW

File: crates/burn-tensor/src/tensor/api/base.rs [hunk 1]
```diff
@@ -171,6 +173,61 @@ where
...
+    /// - If the source and destination dimensions are not of the same length.
+    /// - If the source and destination vectors contain duplicate values.
+    /// - If the source and destination vectors contain values that are out of bounds.
+    ///
+    /// # Returns
+    ///
+    /// The tensor with the dimensions moved.
+    // This is a semantic sugar for `permute`. It is used widely enough, so we define a separate Op
+    // for it
+    pub fn movedim<S: MovedimArgs>(self, src: S, dst: S) -> Tensor<B, D, K> {
```
Comment (LilDojd) [inline crates/burn-tensor/src/tensor/api/base.rs:199]: I've just realised the signature should be more general:

```diff
-    pub fn movedim<S: MovedimArgs>(self, src: S, dst: S) -> Tensor<B, D, K> {
+    pub fn movedim<S1: MovedimArgs, S2: MovedimArgs>(self, src: S1, dst: S2) -> Tensor<B, D, K> {
```

File: crates/burn-tensor/src/tensor/api/base.rs [hunk 2]
```diff
@@ -171,6 +173,61 @@ where
...
+    // for it
+    pub fn movedim<S1: MovedimArgs, S2: MovedimArgs>(self, src: S1, dst: S2) -> Tensor<B, D, K> {
+        let source_dims = src.into_dim_vec::<D>();
+        let destination_dims = dst.into_dim_vec::<D>();
+
+        check!(TensorCheck::movedim_args_length(
+            &source_dims,
+            &destination_dims
+        ));
+        let rank = self.dims().len();
```
Comment (nathanielsimard) [inline crates/burn-tensor/src/tensor/api/base.rs:207]: I think just `D` would work.

File: crates/burn-tensor/src/tensor/api/check.rs [hunk 3]
```diff
@@ -172,6 +173,94 @@ impl TensorCheck {
...
+                TensorError::new("The given dimensions are out of bounds.").details(format!(
+                    "Current tensor has {D} dimensions, but the given dimensions are {:?}.",
+                    dims
+                )),
+            );
+        }
+
+        // Check there are no duplicates
+        let mut uniq = HashSet::new();
+        let duplicates = dims.iter().find(|&&x| !uniq.insert(x));
```
Comment (nathanielsimard) [inline crates/burn-tensor/src/tensor/api/check.rs:228]: I think we should avoid creating hashmap or hashset when possible, especially when working with sequences < 10 in size, just doing for loops would probably be the best here.

File: crates/burn-tensor/src/tensor/api/base.rs [hunk 4]
```diff
@@ -171,6 +173,61 @@ where
...
+
+        let mut m = HashMap::new();
```
Comment (nathanielsimard) [inline crates/burn-tensor/src/tensor/api/base.rs:209]: I would avoid using an hashmap here. maybe a BTreeMap would be lighter.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Thanks for the contribution!

File: crates/burn-tensor/src/tensor/api/base.rs (hunk 2 above)
Comment (LilDojd) [reply crates/burn-tensor/src/tensor/api/base.rs]: Right, my bad

**Comment by LilDojd** [comment]
Thanks for the review! I have addressed your comments in recent commit

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**